src/text/line.rs

```diff
@@ -586,32 +556,100 @@ impl WidgetRef for Line<'_> {
         };
-        for span in &line.spans {
-            let span_width = span.width() as u16;
-            let span_area = Rect {
-                x,
-                width: span_width.min(area.right().saturating_sub(x)),
-                ..area
-            };
-            span.render(span_area, buf);
-            x = x.saturating_add(span_width);
-            if x >= area.right() {
-                break;
+    }
+}
+
+/// Renders all the spans of the line that should be visible. Visibility is determined by the area
+/// and the (remaining) `span_skip_width`.
+#[allow(clippy::cast_possible_truncation, clippy::arithmetic_side_effects)] // Ensured by debug_assert or explained in comment
+fn render_spans(spans: &[Span], mut area: Rect, buf: &mut Buffer, mut span_skip_width: usize) {
+    let spans = spans
+        .iter()
+        .map(|span| (span, span.width()))
+        // Determine whether a span is visible or not. Filter non visible ones out.
+        .filter_map(|(span, span_width)| {
+            // Ignore spans that are completely before the offset. Decrement `span_skip_width` by
+            // the span width until we find a span that is partially or completely visible.
+            if span_skip_width >= span_width {
+                span_skip_width -= span_width;
+                return None;
+            }
+
+            // This applies the skip from the start but not the end.
+            // The end will be trimmed by the actual render into the buffer.
+            // This means that the available_width might be > u16::MAX.
+            let available_width = span_width.saturating_sub(span_skip_width);
+
+            span_skip_width = 0; // ensure that the next span is rendered in full
+
+            Some((span, span_width, available_width))
+        })
+        // Handle truncation when needed
+        .map(|(span, span_width, available_width)| {
+            if span_width > available_width {
+                // Partially visible
+
+                // As the end is truncated by the area width we only need to truncate the start.
+                let (content, actual_width) = span.content.unicode_truncate_start(available_width);
+                // if the truncation has truncated an initial grapheme, then we need to start rendering
+                // from a position that takes that into account by indenting the start of the area
+                let offset_of_initial_grapheme = available_width - actual_width;
+
+                debug_assert!(
+                    u16::try_from(offset_of_initial_grapheme).is_ok(),
+                    "single graphemes are kinda short. Some Emojis reach 22 but not nearly close to u16::MAX"
+                );
+                let offset_of_initial_grapheme = offset_of_initial_grapheme as u16;
+
+                (
+                    content,
+                    span.style,
+                    actual_width,
+                    offset_of_initial_grapheme,
+                )
+            } else {
```

Comment:
This looks a lot better to me! Nice find!

The `if` block still executes just once, right? If so, can we add a comment.

Reply:
Should be. (I don't think we can include a debug_assertion for that one in a neat and simple way?)

But that depends on the `available_width` math from above which is further away. That should be stated in that comment then too.

Reply:
Maybe you can add `.enumerate()` before the `.map()` and `debug_assert!(i == 0)`? LLVM might optimize it away if it sees it is not being used anyway.

Reply:
This is unlikely to ever be a hotspot